libsolidity/analysis/SyntaxChecker.cpp

```diff
@@ -183,6 +186,58 @@ bool SyntaxChecker::visit(Throw const& _throwStatement)
 	return true;
 }
 
+bool SyntaxChecker::visit(Literal const& _literal)
+{
+	// Generic checks no matter what base this number literal is of:
+	if (_literal.value().back() == '_')
+	{
+		m_errorReporter.syntaxError(_literal.location(), "Invalid use of underscores in number literal. No trailing underscores allowed.");
+		return true;
+	}
+
+	if (_literal.value().find("__") != ASTString::npos)
+	{
+		m_errorReporter.syntaxError(_literal.location(), "Invalid use of underscores in number literal. Only one consecutive underscores between digits allowed.");
+		return true;
+	}
+
+	if (!_literal.isHexNumber())
+		// No special validation on grouped decimals yet.
+		return true;
+
+	// We have a hex literal. Do underscore validation
```

Comment:
Do we need this comment at all? It's a bit misleading here, since underscore validation only starts a few lines below, so it should probably at least be moved there.